core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java

```diff
@@ -588,6 +603,115 @@ private Map<NodeRef<Expression>, Type> analyzeExpression(Expression expression)
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    valueSet = ValueSet.ofRanges(Range.lessThan(sourceType, utf8Slice(Integer.toString(date.getYear() + 1))));
+                    break;
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                    valueSet = ValueSet.ofRanges(Range.greaterThan(sourceType, utf8Slice(Integer.toString(date.getYear() - 1))));
+                    break;
+                default:
+                    return Optional.empty();
+            }
+
+            // Date representations starting with whitespace, sign or leading zeroes.
+            valueSet = valueSet.union(ValueSet.ofRanges(
+                    Range.lessThan(sourceType, utf8Slice("1")),
+                    Range.greaterThan(sourceType, utf8Slice("9"))));
+
+            return Optional.of(new ExtractionResult(
+                    TupleDomain.withColumnDomains(ImmutableMap.of(sourceSymbol, Domain.create(valueSet, nullAllowed))),
+                    node));
+        }
+
+        /**
+         * @return Date representations of the form 2005-09-09, 2005-09-9, 2005-9-09 and 2005-9-9 expanded to ranges:
+         * {@code [2005-09-09, 2005-09-0:), [2005-09-9, 2005-09-:), [2005-9-09, 2005-9-0:), [2005-9-9, 2005-9-:)}
+         * (the {@code :} character is the next one after {@code 9}).
+         */
+        private static SortedRangeSet dateStringRanges(LocalDate date, VarcharType domainType)
+        {
+            checkArgument(date.getYear() >= 1000 && date.getYear() <= 9999, "Unsupported date: %s", date);
+
+            int month = date.getMonthValue();
+            int day = date.getDayOfMonth();
+            boolean isMonthSingleDigit = date.getMonthValue() < 10;
+            boolean isDaySingleDigit = date.getDayOfMonth() < 10;
+
+            // A specific date value like 2005-09-10 can be a result of a CAST for number of various forms,
+            // as the value can have optional sign, leading zeros for the year, and surrounding whitespace,
+            // E.g. '  +002005-9-9  '.
+
+            List<Range> valueRanges = new ArrayList<>(4);
+            for (boolean useSingleDigitMonth : List.of(true, false)) {
+                for (boolean useSingleDigitDay : List.of(true, false)) {
+                    if (useSingleDigitMonth && !isMonthSingleDigit) {
+                        continue;
+                    }
+                    if (useSingleDigitDay && !isDaySingleDigit) {
+                        continue;
+                    }
+                    String dateString = date.getYear() +
+                            ((!useSingleDigitMonth && isMonthSingleDigit) ? "-0" : "-") + month +
+                            ((!useSingleDigitDay && isDaySingleDigit) ? "-0" : "-") + day;
+                    String nextStringPrefix = dateString.substring(0, dateString.length() - 1) + (char) (dateString.charAt(dateString.length() - 1) + 1); // cannot overflow
```

Comment:
Sanity question. Char SQL type ordering uses the same order as Java char ordering? ie. if you add one to a char in java you also get the next char when it comes to SQL char sorting

Reply:
In general -- no (i think it breaks for non-BMP)
Here, we're operating within ASCII (digits and hyphen). '9' + 1 produces ':', still ASCII.